src/deluge/gui/views/instrument_clip_view.cpp

```diff
@@ -1302,69 +1332,71 @@ ActionResult InstrumentClipView::padAction(int32_t x, int32_t y, int32_t velocit
-				if (currentPathChars == NULL) {
+				char const* path = afh->filePath.get();
+				if (path == NULL) {
 					continue;
 				}
-				char const* slashAddress = strrchr(currentPathChars, '/');
-				if (slashAddress) {
-					int32_t slashPos = (uint32_t)slashAddress - (uint32_t)currentPathChars;
-					String dir;
-					dir.set(&afh->filePath);
-					dir.shorten(slashPos);
-					FRESULT result = f_opendir(&staticDIR, dir.get());
-					FilePointer thisFilePointer;
-					int32_t numSamples = 0;
-
-					if (result != FR_OK) {
-						display->displayError(ERROR_SD_CARD);
-						return ActionResult::DEALT_WITH;
+				char* slashAddress = strrchr(path, '/');
+				if (slashAddress == NULL) {
+					continue;
+				}
+
+				// Open directory of current audio file
+				*slashAddress = 0;
+				FRESULT result = f_opendir(&staticDIR, path);
+				if (result != FR_OK) {
+					display->displayError(ERROR_SD_CARD);
+					return ActionResult::DEALT_WITH;
+				}
+
+				// Select random audio file from directory
+				int32_t fileCount = 0;
+				while (f_readdir(&staticDIR, &staticFNO) == FR_OK && staticFNO.fname[0] != 0) {
+					audioFileManager.loadAnyEnqueuedClusters();
+					if (staticFNO.fattrib & AM_DIR || !isAudioFilename(staticFNO.fname)) {
+						continue;
 					}
-					while (true) {
-						result = f_readdir_get_filepointer(&staticDIR, &staticFNO,
-						                                   &thisFilePointer); /* Read a directory item */
-						if (result != FR_OK || staticFNO.fname[0] == 0) {
-							break; // Break on error or end of dir
-						}
-						if (staticFNO.fname[0] == '.' || staticFNO.fattrib & AM_DIR
-						    || !isAudioFilename(staticFNO.fname)) {
-							continue; // Ignore dot entry
-						}
-						audioFileManager.loadAnyEnqueuedClusters();
-						fnArray[numSamples].set(staticFNO.fname);
-						numSamples++;
-						if (numSamples >= MaxFiles) {
-							break;
-						}
+					if (random(fileCount++) == 0) { // Algorithm: Reservoir Sampling with k=1
+						strncpy(chosenFilename, staticFNO.fname, 256);
 					}
+				}
 
-					if (numSamples >= 2) {
-						soundDrum->unassignAllVoices();
-						afh->setAudioFile(NULL);
-						String filePath; //add slash
-						filePath.set(&dir);
-						int32_t dirWithSlashLength = filePath.getLength();
-						if (dirWithSlashLength) {
-							filePath.concatenateAtPos("/", dirWithSlashLength);
-							dirWithSlashLength++;
-						}
-						char const* fn = fnArray[random(numSamples - 1)].get();
-						filePath.concatenateAtPos(fn, dirWithSlashLength);
-						AudioEngine::stopAnyPreviewing();
-						afh->filePath.set(&filePath);
-						afh->loadFile(false, true, true, 1, 0, false);
-						soundDrum->name.set(fn);
-						numRandomized++;
-						getCurrentInstrument()->beenEdited();
-					}
+				// Assign new audio file
+				if (fileCount) {
+					String filePath;
+					filePath.set(path);
+					filePath.concatenate("/");
+					filePath.concatenate(chosenFilename);
```

Comment:
Rather than allocating a new string, I think we can re-use the one at `afh->filePath`

Reply:
Done!